### PR TITLE
feat(filer): lazy metadata pulling

### PR DESCRIPTION
**Pull request by petedodd-pd**
# What problem are we solving?
Part of #8391.

SeaweedFS has no mechanism to lazily pull object metadata from remote storage on demand. This PR implements lazy metadata pull at the filer layer so that any filer client (S3, FUSE, etc.) benefits without per-client logic.

# How are we solving the problem?
When `FindEntry` misses in the filer store, the filer checks whether the path is under a remote-storage mount. If so, it stats the remote via `StatFile`, builds a filer entry from the result, and persists it via `CreateEntry` with `SkipCheckParentDirectory` so parent directories under the mount are not required.

- **Placement in filer** — Lazy fetch is implemented in `maybeLazyFetchFromRemote`, invoked from `FindEntry` on a store miss. All filer clients (S3, FUSE, etc.) get lazy metadata pull automatically.
- **Deduplication** — Each `Filer` instance has its own `singleflight.Group` to deduplicate concurrent lookups for the same path.
- **Recursion guard** — A context key prevents re-entry when `CreateEntry` triggers `FindEntry` again.
- **Availability over consistency** — If `CreateEntry` fails after a successful `StatFile`, the in-memory entry is still returned; the singleflight key is forgotten so the next lookup retries the filer write.
- **Path handling** — `relPath` is normalised to include a leading slash when `mountDir` is root, so remote paths are built correctly.

Uses the existing `FilerRemoteStorage` (`FindMountDirectory`, `FindRemoteStorageClient`). Longest-prefix mount resolution is preserved via ptrie iteration order.

# How is the PR tested?
Unit tests covering:

- `maybeLazyFetchFromRemote`: hits remote and persists; not under mount; remote object not found; create entry failure returns in-memory entry; longest-prefix mount wins; context guard prevents recursion
- `FindEntry` lazy fetch on store miss, with second lookup served from store
- `FindMountDirectory` longest-prefix semantics
- `registerStubMaker` restores the previous maker on cleanup

# Checks
- [x] I have added unit tests if possible.
- [ ] I will add related wiki document changes and link to this PR after merging.
- [x] All AI code review comments have been addressed. No more comments to fix if reviewed again. Reviewer may request additional gemini and copilot reviews.

# Checks for AI generated PRs
- [x] I have reviewed every line of code.
- [x] The PR is kept as minimum as possible. Large PRs would not be accepted.

## Summary by CodeRabbit

* **New Features**
  * Lazily fetches missing file entries from remote storage on-demand, with single-request deduplication, recursion protection, timeouted persistence, and graceful handling when remote objects are absent or persistence fails.
  * Remote mounts resolve using the longest matching mount prefix for correct multi-mount behavior.

* **Tests**
  * Extensive coverage of lazy-fetch flow, caching behavior, mount-prefix resolution, recursion guards, persistence edge cases, and error handling.

### CODE REVIEW

File: weed/s3api/s3api_remote_storage_index.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+			SkipCheckParentDirectory: true,
+		})
+		return createErr
+	})
+	if saveErr != nil {
+		glog.Warningf("lazyFetchFromRemote: failed to persist filer entry for %s/%s: %v", bucket, object, saveErr)
+		s3a.remoteStorageIdx.fetchGroup.Forget(objectFilerPath)
+	}
+
+	return entry, nil
```
[the remark this answers is not in the document]
Comment (petedodd-pd) [reply weed/s3api/s3api_remote_storage_index.go]: The comment is incorrect. `entry` is built from a successful `StatFile` — the remote metadata is valid regardless of whether the filer write succeeded. The filer write is intentionally best-effort; propagating `saveErr` would return a 500 to the client for a perfectly accessible remote object on any transient filer hiccup. No change needed.

File: weed/s3api/s3api_remote_storage_index.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+	objectFilerPath := bucketFilerPath + "/" + strings.TrimPrefix(object, "/")
+
+	client, mountLoc, relPath, found := s3a.remoteStorageIdx.findForPath(objectFilerPath)
+	if !found {
+		return nil, nil
+	}
+
+	result, err, _ := s3a.remoteStorageIdx.fetchGroup.Do(objectFilerPath, func() (interface{}, error) {
+		return s3a.doLazyFetch(context.WithoutCancel(ctx), objectFilerPath, client, mountLoc, relPath, bucket, object)
+	})
```
[the remark this answers is not in the document]
Comment (petedodd-pd) [reply weed/s3api/s3api_remote_storage_index.go]: `StatFile` on the S3 client takes no context — the interface signature is `StatFile(loc *remote_pb.RemoteStorageLocation)` with no `context.Context` parameter. Applying a timeout in `doLazyFetch` would have no effect since the context is never passed down to `HeadObject`. The fix would require changing the `RemoteStorageClient` interface and all its implementations — that's a separate, larger PR touching multiple packages. No change needed here.

File: weed/s3api/s3api_server.go [hunk 3]
```diff
@@ -80,6 +80,7 @@ type S3ApiServer struct {
 	embeddedIam           *EmbeddedIamApi // Embedded IAM API server (when enabled)
 	stsHandlers           *STSHandlers    // STS HTTP handlers for AssumeRoleWithWebIdentity
 	cipher                bool            // encrypt data on volume servers
+	remoteStorageIdx      *remoteStorageIndex
```
Comment (chrislusf) [inline weed/s3api/s3api_server.go]: metadata operation is managed by filer, not in s3

**Comment by petedodd-pd** [comment]
@coderabbitai resume

**Comment by petedodd-pd** [comment]
@coderabbitai resume

**Comment by chrislusf** [comment]
When will the pulled metadata be invalidated?

**Comment by petedodd-pd** [comment]
@chrislusf currently, only on a deletes, as this will delete the local and remote data for that entry (if it exists remotely).

I believe currently the setup for a remote storage bucket assumes the metadata doesn't change from the remote side as metadata is only pulled all at once upon mounting. Therefore, I'm not sure if this needs invalidation. I think if your bucket/directory is mounted remotely, it would always be desirable to pull the metadata for a file if it doesn't exist in SeaweedFS but exists remotely.

If there is a fear of accruing costs due to checking the remote bucket on each metadata miss on seaweed, then we can always hide this feature behind a flag. As long as the full lazy metadata approach is documented in the wiki so users don't get confused, this should be fine.

**Comment by chrislusf** [comment]
we can use this to uncache:
```
	remote.uncache -minCacheAge=3600   # uncache files older than 1 hour (cached time)
```
